samples/extensions/sparse_image/sparse_image.cpp

```diff
@@ -0,0 +1,1430 @@
+	subresource_range.levelCount     = 1U;
+
+	if (current_mip_level != 0xFF && current_mip_level != 0U)
+	{
+		subresource_range.baseMipLevel = current_mip_level;
+		vkb::image_layout_transition(command_buffer, virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, subresource_range);
+
+		subresource_range.baseMipLevel = current_mip_level - 1U;
+		vkb::image_layout_transition(command_buffer, virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, subresource_range);
+	}
+	else if (current_mip_level == 0U)
+	{
+		subresource_range.baseMipLevel = current_mip_level;
+		vkb::image_layout_transition(command_buffer, virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, subresource_range);
+	}
+
+	device->flush_command_buffer(command_buffer, queue, true);
+}
+
+void SparseImage::free_unused_memory()
+{
+	for (size_t i = 0U; i < virtual_texture.page_table.size(); i++)
+	{
+		virtual_texture.page_table[i].gen_mip_required = false;
+
+		if (virtual_texture.page_table[i].render_required_set.empty() && virtual_texture.page_table[i].valid)
+		{
+			virtual_texture.page_table[i].valid = false;
+			virtual_texture.available_memory_page_indices.push_back(virtual_texture.page_table[i].page_index);
+		}
+	}
+}
+
+void SparseImage::load_least_detailed_level()
+{
+	set_least_detailed_level();
+	compare_mips_table();
+	while (!virtual_texture.texture_block_update_set.empty())
+	{
+		process_texture_blocks();
+		update_and_generate();
+		free_unused_memory();
+	}
+}
+
+void SparseImage::process_stage(enum Stages next_stage)
+{
+	switch (next_stage)
+	{
+		case SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE:
+			calculate_mips_table(current_mvp_transform, on_screen_num_vertical_blocks, on_screen_num_horizontal_blocks, virtual_texture.new_mip_table);
+			this->next_stage = SPARSE_IMAGE_COMPARE_MIPS_TABLE_STAGE;
+			break;
+
+		case SPARSE_IMAGE_COMPARE_MIPS_TABLE_STAGE:
+			compare_mips_table();
+			if (virtual_texture.update_required)
+			{
+				this->next_stage = SPARSE_IMAGE_PROCESS_TEXTURE_BLOCKS_STAGE;
+			}
+			else
+			{
+				this->next_stage = SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE;
+			}
+			frame_counter = 0U;
+			break;
+
+		case SPARSE_IMAGE_PROCESS_TEXTURE_BLOCKS_STAGE:
+			process_texture_blocks();
+			this->next_stage = SPARSE_IMAGE_UPDATE_AND_GENERATE_STAGE;
+			break;
+
+		case SPARSE_IMAGE_UPDATE_AND_GENERATE_STAGE:
+			update_and_generate();
+			free_unused_memory();
+
+			if (frame_counter > 10U)
+			{
+				this->next_stage = SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE;
+			}
+			else if (!transfer_finished)
+			{
+				this->next_stage = SPARSE_IMAGE_PROCESS_TEXTURE_BLOCKS_STAGE;
+			}
+			else
+			{
+				virtual_texture.update_required = false;
+				this->next_stage                = SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE;
+			}
+			break;
+
+		default:
+			break;
```

Comment:
Does this switch catch all cases? Then you could add some `assert(false);`, or so, to "document" that.